### PR TITLE
DAOS-11703 test: Add Catastrohpic Recovery Pass 0 tests

### PR DESCRIPTION
**Pull request by shimizukko**
DAOS-11703: Checker can start with either AdminExcluded or CheckerStarted.
Test checker can only be run when the system status is AdminExcluded.
1. Call dmg check enable.
2. Stop rank 1.
3. Call dmg system query and check that status of at least one rank is not "checkerstarted". We verify that the new checkerstarted state is properly changed.
4. Call dmg check start. It should show error because the stopped rank is not at CheckerStarted state.
5. Call dmg check query. It should show error because the stopped rank is not at CheckerStarted state.
6. Call dmg check disable. It should work.
7. Restart the stopped rank for cleanup.

DAOS-11704: Admin can set rank state as AdminExcluded during crash.
Test admin can enable and disable the rank state to AdminExcluded when the rank is down.
1. Stop rank 1.
2. Set rank 1 to AdminExcluded state by calling dmg system exclude --ranks=1.
3. Verify that the rank 1 state is AdminExluded by calling dmg system query.
4. Verify that the checker can be run with AdminExcluded state by calling enable, start, query, and disable. Verify that none of the commands returns error.
5. Disable AdminExcluded by calling dmg system clear-exclude --ranks=1
6. Verify that the rank 1 state is Excluded.
7. Start rank 1 for clean up. (At this point, its state is "Excluded".)

Skip-unit-tests: true
Skip-fault-injection-test: true
Test-tag: pr test_container_basics pass_0
Required-githooks: true

Signed-off-by: Makito Kano <makito.kano@intel.com>

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

File: src/tests/ftest/recovery/pass_0.py [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+        # changed.
+        query_out = dmg_command.system_query()
+        not_checkerstarted_found = False
+        for member in query_out["response"]["members"]:
+            if member["state"] != "checkerstarted":
+                not_checkerstarted_found = True
+                break
+        if not not_checkerstarted_found:
+            # All rank's status is "checkerstarted".
+            errors.append("All rank's status is checkerstarted!")
```
Comment (daltonbohning) [inline src/tests/ftest/recovery/pass_0.py:59]: Would it be better to fail immediately?

File: src/tests/ftest/util/dmg_utils.py [hunk 2]
```diff
@@ -1295,6 +1325,41 @@ def version(self):
         """
         return self._get_result(["version"])
 
+    def check_enable(self, pool=None):
+        """Call dmg check enable.
+
+        Args:
+            pool
```
Comment (daltonbohning) [inline src/tests/ftest/util/dmg_utils.py:1332]: docstring?

File: src/tests/ftest/util/dmg_utils.py [hunk 3]
```diff
@@ -1295,6 +1325,41 @@ def version(self):
...
+        """
+        return self._get_json_result(("check", "enable"), pool=pool)
+
+    def check_start(self, pool=None, dry_run=False, reset=False, failout=False,
+                    auto=False):
+        """Call dmg check start.
+
+        Args:
+            pool
```
Comment (daltonbohning) [inline src/tests/ftest/util/dmg_utils.py:1341]: docstring?

File: src/tests/ftest/util/dmg_utils.py [hunk 4]
```diff
@@ -1295,6 +1325,41 @@ def version(self):
...
+        """
+        return self._get_json_result(
+            ("check", "start"), pool=pool, dry_run=dry_run, reset=reset, failout=failout,
+            auto=auto)
+
+    def check_query(self, pool=None):
+        """Call dmg check query.
+
+        Args:
+            pool
```
Comment (daltonbohning) [inline src/tests/ftest/util/dmg_utils.py:1351]: docstring?

File: src/tests/ftest/util/dmg_utils.py [hunk 5]
```diff
@@ -1295,6 +1325,41 @@ def version(self):
...
+        """
+        return self._get_json_result(("check", "query"), pool=pool)
+
+    def check_disable(self, pool=None):
+        """Call dmg check disable.
+
+        Args:
+            pool
```
Comment (daltonbohning) [inline src/tests/ftest/util/dmg_utils.py:1359]: docstring?

**Review by daltonbohning — COMMENTED** [review]
Just minor comments

File: src/tests/ftest/recovery/pass_0.py (hunk 1 above)
Comment (shimizukko) [reply src/tests/ftest/recovery/pass_0.py]: Yes. If all rank's status is `checkerstarted`, it'll affect the next steps, so it's better to fail here immediately. Fixed.

File: src/tests/ftest/util/dmg_utils.py (hunk 2 above)
Comment (shimizukko) [reply src/tests/ftest/util/dmg_utils.py]: Added

File: src/tests/ftest/util/dmg_utils.py (hunk 3 above)
Comment (shimizukko) [reply src/tests/ftest/util/dmg_utils.py]: Added

File: src/tests/ftest/util/dmg_utils.py (hunk 4 above)
Comment (shimizukko) [reply src/tests/ftest/util/dmg_utils.py]: Added

File: src/tests/ftest/util/dmg_utils.py (hunk 5 above)
Comment (shimizukko) [reply src/tests/ftest/util/dmg_utils.py]: Added

File: src/tests/ftest/recovery/pass_0.py [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+        7. Restart the stopped rank for cleanup.
+
+        Jira ID: DAOS-11703
+
+        :avocado: tags=all,weekly_regression
+        :avocado: tags=vm
+        :avocado: tags=recovery,pass_0
+        :avocado: tags=test_checker_on_admin_excluded
+        """
+        errors = []
```
Comment (daltonbohning) [inline src/tests/ftest/recovery/pass_0.py:39]: Now this isn't used at all :)
```suggestion
```

File: src/tests/ftest/recovery/pass_0.py [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+            self.log.info("dmg check query is expected to fail. Error: %s", error)
+
+        # 6. Call dmg check disable. It should work.
+        dmg_command.check_disable()
+
+        # 7. Restart the stopped rank for cleanup.
+        self.log.info("Restart stopped rank for cleanup.")
+        dmg_command.system_start(ranks="1")
+
+        report_errors(test=self, errors=errors)
```
Comment (daltonbohning) [inline src/tests/ftest/recovery/pass_0.py:82]:
```suggestion
```

File: src/tests/ftest/recovery/pass_0.py [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+        admin_excluded_found = False
+        rank_1_state = None
+        for member in query_out["response"]["members"]:
+            if member["rank"] == 1:
+                rank_1_state = member["state"]
+                if rank_1_state == "adminexcluded":
+                    admin_excluded_found = True
+                break
+        if not admin_excluded_found:
+            errors.append("Rank 1 state is not AdminExcluded! {}".format(rank_1_state))
```
Comment (daltonbohning) [inline src/tests/ftest/recovery/pass_0.py:124]: Similar to above test, do we want to fail immediately?

File: src/tests/ftest/recovery/pass_0.py (hunk 6 above)
Comment (shimizukko) [reply src/tests/ftest/recovery/pass_0.py]: Removed. Thanks

File: src/tests/ftest/recovery/pass_0.py (hunk 7 above)
Comment (shimizukko) [reply src/tests/ftest/recovery/pass_0.py]: Removed

File: src/tests/ftest/recovery/pass_0.py (hunk 8 above)
Comment (shimizukko) [reply src/tests/ftest/recovery/pass_0.py]: Yes. Updated to call `self.fail`.

**Review by daltonbohning — APPROVED** [review]

File: src/tests/ftest/util/dmg_utils.py [hunk 9]
```diff
@@ -1048,6 +1063,21 @@ def system_erase(self):
         """
         return self._get_json_result(("system", "erase"))
 
+    def system_exclude(self, ranks):
```
Comment (rpadma2) [inline src/tests/ftest/util/dmg_utils.py:1066]: Any reason why we don't support rank-hosts parameter?

File: src/tests/ftest/util/dmg_utils.py [hunk 10]
```diff
@@ -965,6 +965,21 @@ def system_cleanup(self, machinename=None, verbose=True):
         return self._get_json_result(
             ("system", "cleanup"), machinename=machinename, verbose=verbose)
 
+    def system_clear_exclude(self, ranks):
```
Comment (rpadma2) [inline src/tests/ftest/util/dmg_utils.py:968]: Any reason why we don't support rank-hosts parameter?

File: src/tests/ftest/recovery/pass_0.yaml [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+
+timeout: 180
+
+server_config:
+  name: daos_server
+  engines_per_host: 1
+  engines:
+    0:
+      targets: 1
+      nr_xs_helpers: 0
```
Comment (rpadma2) [inline src/tests/ftest/recovery/pass_0.yaml:12]: Any reason why nr_xs_helpers is set to 0. Can we set to 1... Also, why targets: 1... Why not use the default targets?

**Review by rpadma2 — APPROVED** [review]

**Pull request merged**